### PR TITLE
Serverless function

### PR DESCRIPTION
**Pull request by bardabun**
I've made several improvements to the 'azure-serverless' project, enhancing log shipping from Azure to Logz.io. Key updates include:

1. Refined deployment steps for easier setup.
2. Improved configuration template with clearer parameters.
3. Detailed instructions for streaming Azure service data.
4. New backup mechanism for unshipped logs.
5. Updated documentation with helpful visuals.

These changes aim to streamline user experience and boost functionality. Looking forward to your feedback.

### CODE REVIEW

File: LogzioShipper/__init__.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+# Initialize Application Insights
+tc = TelemetryClient(os.getenv("APPINSIGHTS_INSTRUMENTATIONKEY"))
+
+# Load environment variables from a .env file for local development
+load_dotenv()
+
+# Initialize Azure Blob Storage container client
+container_client = ContainerClient.from_connection_string(
+    conn_str=os.getenv("AzureWebJobsStorage"),  # On Azure
+    # conn_str=os.getenv("AZURE_STORAGE_CONNECTION_STRING"),  # On local
```
Comment (mirii1994) [inline LogzioShipper/__init__.py:28]: Please remove commented out line if it's unnecessary :)

File: LogzioShipper/__init__.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    container_name=os.getenv("AZURE_STORAGE_CONTAINER_NAME")
+)
+
+# Configure logging
+logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)s %(message)s')
+# logging.getLogger().addHandler(logging.StreamHandler(sys.stdout))
```
Comment (mirii1994) [inline LogzioShipper/__init__.py:34]: Please remove commented out line if it's unnecessary :)

File: LogzioShipper/__init__.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+
+# Logz.io configuration
+# LOGZIO_URL = os.getenv("LOGZIO_LISTENER")
```
Comment (mirii1994) [inline LogzioShipper/__init__.py:38]: Please remove commented out line if it's unnecessary :)

File: LogzioShipper/__init__.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+LOGZIO_URL = os.getenv("LogzioURL")
+# LOGZIO_TOKEN = os.getenv("LOGZIO_TOKEN")
```
Comment (mirii1994) [inline LogzioShipper/__init__.py:40]: Please remove commented out line if it's unnecessary :)

File: LogzioShipper/__init__.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+def delete_empty_fields_of_log(log):
+    if isinstance(log, dict):
+        return {k: v for k, v in log.items() if v is not None and v != ""}
+    elif isinstance(log, list):
+        return [delete_empty_fields_of_log(item) for item in log]
+    else:
+        return log
+
+
+@backoff.on_exception(backoff.expo, requests.exceptions.RequestException, max_tries=3)
```
Comment (mirii1994) [inline LogzioShipper/__init__.py:75]: Can we make the max_tries configurable?

File: LogzioShipper/__init__.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+def send_batch(batch_data):
+    try:
+        batch_str = ''.join(batch_data)
+        response = session.post(LOGZIO_URL, params={"token": LOGZIO_TOKEN, "type": "eventHub"}, headers=HEADERS, data=batch_str)
```
Comment (mirii1994) [inline LogzioShipper/__init__.py:79]: Let's allow the type to be configurable, with default value of eventHub

File: LogzioShipper/__init__.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    for event in azeventhub:
+        logs = process_eventhub_message(event)
+        for log in logs:
+            formatted_log = json.dumps(log) + '\n'
+            local_log_batch.append(formatted_log)
+
+        # Check the batch size or time interval after processing all logs from the event
+        current_time = time.time()
+        if len(local_log_batch) >= BUFFER_SIZE or (current_time - last_batch_time) >= INTERVAL_TIME:
+            print(f"Adding batch of size {len(local_log_batch)} to the sending queue.")
```
Comment (mirii1994) [inline LogzioShipper/__init__.py:101]: If there's a logger let's use it throughout the code, instead of prints

File: README.md [hunk 8]
```diff
@@ -11,7 +11,7 @@ This repository contains the Python code and instructions necessary to ship logs
```
Comment (mirii1994) [inline README.md:11]: Please rech out to @nico-shishkin for him to review the readme :)

File: deployments/azuredeploylogs.json [hunk 9]
```diff
@@ -0,0 +1,307 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters":
+    {
+        "ContainerName":
+        {
+            "defaultValue": "logs",
```
Comment (mirii1994) [inline deployments/azuredeploylogs.json:8]: Let's change it to a name that is more clear on the purpose of the container - if it's for backup logs, then something like "backup logs"

File: deployments/azuredeploylogs.json [hunk 10]
```diff
@@ -0,0 +1,307 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters":
+    {
+        "ContainerName":
```
Comment (mirii1994) [inline deployments/azuredeploylogs.json:6]: Let's change the parameter name to a name that is more clear on the purpose of the container - if it's for backup logs, then something like "backup logs"

File: deployments/azuredeploylogs.json [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+                "https://listener-au.logz.io:8071",
+                "https://listener-ca.logz.io:8071",
+                "https://listener-eu.logz.io:8071",
+                "https://listener-nl.logz.io:8071",
+                "https://listener-uk.logz.io:8071",
+                "https://listener-wa.logz.io:8071"
+            ],
+            "metadata":
+            {
+                "description": "The Logz.io listener URL for your region."
```
Comment (mirii1994) [inline deployments/azuredeploylogs.json:33]: Please reach out to @nico-shishkin to go over the descriptions of that parameters

File: deployments/azuredeploylogs.tf [hunk 12]
```diff
@@ -0,0 +1,118 @@
+provider "azurerm" {
```
Comment (mirii1994) [inline deployments/azuredeploylogs.tf:1]: If we include deployment via Terraform, we need to include instructions for this method in our readme

**Review by mirii1994 — CHANGES_REQUESTED** [review]
Great great work!!!
Please see my comments :)

File: README.md [hunk 13]
```diff
@@ -55,22 +95,16 @@ Allow some time for data to flow from Azure to Logz.io, then check your Logz.io
...
 ### Post-Deployment Configuration
 
 To modify configuration after deployment, visit your Function App's 'Configuration' tab. You can adjust settings such as LogzioURL, LogzioToken, bufferSize, and more.
 
-![Function's Configuration](img/configuration-settings-python.png)
-
 ## Changelog
 
-- Version 1.0.0:
+- Version 0.0.1:
```
Comment (mirii1994) [inline README.md:106]:
```suggestion
- **0.0.1**:
```

**Review by mirii1994 — APPROVED** [review]
Great work!!!!
Let's wait for Nico's review of the documentation, and then we'll proceed :)

**Review by nico-shishkin — APPROVED** [review]

**Pull request merged**